### PR TITLE
Authed endpoint for retrieving list

### LINKED ISSUE #123: List management tooling: Generate new policy.json

### PR DESCRIPTION
**Pull request by sydneyli**
fixes #123 
`/auth/list` works and has been tested ~now need to finish testing `/auth/sync`.~
EDIT: Decided to fix this up to only include the `/auth/list` endpoint. `/auth/sync` will come in a future PR.
Sub-part of #6

### CODE REVIEW

File: list_api.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+//       expire_weeks: after how many weeks should this list expire? If unset
+//                     or invalid, defaults to 2. If set to 0, expires immediately.
+//       queued_weeks: for at least how many weeks should domains on the resulting list
+//                     have been queued? if unset or invalid, defaults to 1.
+func (api API) GetList(r *http.Request) APIResponse {
+	expireWeeks := getNumberParam(r, "expire_weeks", 2)
+	queuedWeeks := getNumberParam(r, "queued_weeks", 1)
+	list := api.List.Raw()
+	list.Timestamp = time.Now()
+	list.Expires = list.Timestamp.Add(time.Hour * 24 * 7 * time.Duration(expireWeeks))
```
Comment (vbrown608) [inline list_api.go]: I'm not sure I understand what `list.Expires` is doing here. Is it just returning the current time plus the `expireWeeks` param to the user as JSON?

File: policy/policy.go [hunk 2]
```diff
@@ -97,21 +102,69 @@ func (l UpdatedList) Get(domain string) (TLSPolicy, error) {
...
+
+func (p Pinset) clone() Pinset {
+	pinset := Pinset{StaticSPKIHashes: make([]string, 0)}
+	for _, hash := range p.StaticSPKIHashes {
+		pinset.StaticSPKIHashes = append(pinset.StaticSPKIHashes, hash)
+	}
+	return pinset
+}
+
+func (p TLSPolicy) clone() TLSPolicy {
```
Comment (vbrown608) [inline policy/policy.go]: Wow, making a deep copy in Go is not very straightfoward! I agree we need to manually clone the nested structs/slices.

By my reading of this [blog post](https://flaviocopes.com/go-copying-structs/#deep-copy-a-struct), it seems that we could pare this down a bit by taking advantage of the fact that Go copies basic types by value when a struct is assigned to a new variable. That would apply to basic types in `list` as well.

File: list_api.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+//   GET /auth/list
+//       expire_weeks: after how many weeks should this list expire? If unset
+//                     or invalid, defaults to 2. If set to 0, expires immediately.
+//       queued_weeks: for at least how many weeks should domains on the resulting list
+//                     have been queued? if unset or invalid, defaults to 1.
+func (api API) GetList(r *http.Request) APIResponse {
+	expireWeeks := getNumberParam(r, "expire_weeks", 2)
+	queuedWeeks := getNumberParam(r, "queued_weeks", 1)
+	list := api.List.Raw()
+	list.Timestamp = time.Now()
```
[the remark this answers is not in the document]
Comment (sydneyli) [reply list_api.go]: Hmm, I couldn't find this! This is primarily to update the timestamp of this copy.

File: list_api.go (hunk 1 above)
Comment (sydneyli) [reply list_api.go]: Yep! To more properly motivate this endpoint, every week or so we release an update to the policy list, and it should have an updated timestamp/expiry date. These timestamps guarantee freshness to consumers of the list, as we can't provide many guarantees once the expiry date has passed.

So every week, in order to generate the "new" list, I've been using this branch to 1) update the timestamp to now and 2) set the expiry date to some time in the future. [Here's a sample](https://github.com/EFForg/starttls-everywhere/blob/master/RULES.md) of what the policy file looks like.

Comment (vbrown608) [reply list_api.go]: Got it! I think when I reviewed this I didn't understand that it was meant to produce the actual list that will be distributed, and not just a convenient summary for you to do stuff with.

File: policy/policy_test.go [hunk 4]
```diff
@@ -81,12 +81,28 @@ func TestHostnamesForDomain(t *testing.T) {
...
+	returned, err := list.HostnamesForDomain("eff.org")
 	if err != nil {
 		t.Fatalf("Encountered %v", err)
 	}
 	if !reflect.DeepEqual(returned, hostnames) {
 		t.Errorf("Expected %s, got %s", hostnames, returned)
 	}
 }
+
+func TestCloneDoesntChangeOriginal(t *testing.T) {
```
Comment (vbrown608) [inline policy/policy_test.go]: :bowtie:

File: list_api.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (vbrown608) [reply list_api.go]: Oh, yeah, I must have been a bit confused. I guess the timestamp could be set in `UpdatedList.Raw()`, but that might confuse the usefulness of that function as strictly a copy. This seems better!

**Pull request merged**